### PR TITLE
feat(metrics): Implement redis blaster support for the cardinality limiter [SNS-1752]

### PR DESCRIPTION
**Pull request by untitaker**
## Background

We are in the midst of rolling out the cardinality limiter to production and
have learned two things since then:

* The load does not fit onto a single redis node
* We would like to use Google's "memorystore", i.e. managed Redis, which does
  not support Redis clusters.

So, we want to implement basic application sharding of keys, where keys are
hashed into slots and routed to individual redis clients.

Initially I set out to implement this straight inside of the cardinality
limiter. However, while easy to build, the resulting code turned out to be
quite messy.

So, against what we discussed, I decided to use RB.

We already support redis blaster for some of our redis-based services
(projectconfig debounce cache eg), it even comes with niceties like
multithreading when talking to multiple Redis instances (whereas in our manual
implementation we'd just have taken the perf hit i guess), and the code changes
to support it in cardinality limiter turned out to be significantly less effort.

### CODE REVIEW

**Comment by mwarkentin** [comment]
🤞🏻 that this is resolved with native redis-cluster support in Memorystore in the next 3-6 months and we can switch over to that for high throughput cases.

Thanks for putting in the work to let us do client-side sharding for now!

File: src/sentry/ratelimits/cardinality.py
Comment (nikhars) [inline src/sentry/ratelimits/cardinality.py:298]: I am wondering if it would be better the implementations based on redis cluster and redis blaster be in separate functions for both check_within_quotas and use_quotas. Then during init you can setup the appropriate function to be used based on the appropriate backend. Then in this code you just call those function callbacks. It makes reasoning about the code a bit simpler.

File: src/sentry/ratelimits/cardinality.py
Comment (onewland) [inline src/sentry/ratelimits/cardinality.py:232]: what is arg 1 (`""`) here?

File: src/sentry/ratelimits/cardinality.py
Comment (onewland) [inline src/sentry/ratelimits/cardinality.py:310]: I'm not familiar with redis-blaster. What is `map()` doing here?

File: src/sentry/ratelimits/cardinality.py
Comment (onewland) [inline src/sentry/ratelimits/cardinality.py:397]: This might be fine but should a layer be built to encapsulate the cluster/rb DAOs? `pipeline()` and `map()` might return similar objects (they both accept `setex/sadd`) but clearly not the same type and I'd be concerned about introducing errors with this logic.

It seems as if these can behave a bit differently. Presumably nothing is happening in `self.client.pipeline(transaction=False)` until `execute()` is called, but `self.client.map()` is executing the redis commands as it goes.

File: tests/sentry/ratelimits/test_cardinality.py
Comment (onewland) [inline tests/sentry/ratelimits/test_cardinality.py:14]: I've never seen this automatic parameter-ization with pytest before -- it's nice.

**Review by onewland — COMMENTED** [review]
I have some questions but they shouldn't block merge. I'm unfamiliar with Rb (redis blaster) so it's possible my comments are missing relevant context.

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py]: @onewland `ctx_mgr` should've been typed `Union[rb.MappingClient, rediscluster.Something]` so i hoped this was typesafe (method access can only happen if the method exists on both union variants), but after getting legibility feedback from nikhar i pulled in another abstraction anyway. I think that's what you meant with DAO too?

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py]: i pulled out this code into a helper class with a somewhat typed interface.

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py:234]: `get_dynamic_cluster_from_options` assumes the caller knows which sentry setting name is being used to configure this. based on that it emits some diagnostics if the settings are invalid. this can't apply here because we're configuring this cardinality limiter through multiple config layers

File: src/sentry/ratelimits/cardinality.py
Comment (onewland) [reply src/sentry/ratelimits/cardinality.py]: I was thinking something slightly different but this addressed my concern

**Review by onewland — APPROVED** [review]

File: src/sentry/ratelimits/cardinality.py
Comment (fpacifici) [inline src/sentry/ratelimits/cardinality.py:250]: Nobody seems to use this method.

**Review by fpacifici — COMMENTED** [review]
Please address the comments before merging.

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py:250]: it's a method on `Service`, but since we don't use `LazyServiceWrapper` (which does call it) it's fair to say it doesn't matter whether we implement it. Still I think we should implement the interface properly.

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py]: fixed

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py]: i actually went with removing this method to reduce the number of attributes that need to be stored.

File: src/sentry/ratelimits/cardinality.py
Comment (untitaker) [reply src/sentry/ratelimits/cardinality.py]: also yes i tested manually (using multiple `db`)

**Pull request merged**
